Organize and add events and add inventory types

AL1L opened the pull request.

AL1L committed:
Move around events

AL1L committed:
Move around events

AL1L committed:
Move all event type exports to one file

AL1L committed:
Move all event type exports to one file

connerdouglass commented:
I really like this improvement! It might sound pedantic, but could we name the folders "block" and "entity", instead of "blocks" and "entities".

Also make sure to run `npm run build:check` to make sure all the other imports still point to the right place. I think there's one in `src/entity/Entity.ts` that needs updating.

Then I will merge it!

AL1L committed:
Chunk events

AL1L committed:
Chunk events

AL1L committed:
Start inventory events

AL1L committed:
Start inventory events

AL1L committed:
Cancellable interface

Created this to make sure events that are cancellable have the required methods and documentation doesnt have to be repeated over and over

AL1L committed:
Cancellable interface

Created this to make sure events that are cancellable have the required methods and documentation doesnt have to be repeated over and over

AL1L committed:
Import inventory events

AL1L committed:
Import inventory events

GaetanLBH committed:
Add files via upload

added the entity spawn event

AL1L committed:
Merge remote-tracking branch 'origin/events' into events

AL1L committed:
Inventory types and events

AL1L committed:
Add missing getBukkitClasspath's

AL1L commented on src/events/EventResult.ts:
```diff
@@ -0,0 +1,5 @@
+export enum EventResult {
```
Maybe instead of resolving, use strings like other enums

@connerdouglass thoughts?

connerdouglass replied:
Yes I think we should use strings in the enum. I'll also comment on `InventoryInteractEvent` with how we should update that.

AL1L replied:
Changed to be consistent with the rest of the project

AL1L commented on src/events/types/block/BlockBreakEvent.ts:
```diff
@@ -28,18 +29,10 @@ export class BlockBreakEvent extends BlockExpEvent {
-    /**
-     * Sets whether or not this block will drop items
-     * @param drop_items the drop items state of this event
-     */
```
Why were these docs removed?

AL1L commented on src/events/types/inventory/EnchantmentsToAdd.ts:
```diff
@@ -0,0 +1,74 @@
+  public whichButton(): number {
+    return this.toJava().whichButton();
+  }
+}
```
Add eof newline

AL1L commented on src/events/types/inventory/EnchantmentsToAdd.ts:
```diff
@@ -0,0 +1,74 @@
+  //  */
+  // public getEnchantsToAdd(): Map<Enchantment, number> {
+  //   return this.toJava().getEnchantsToAdd();
+  // }
```
@connerdouglass I do need some help on this one

connerdouglass replied:
I think we should skip this method for this PR, and then open an Issue to implement it separately.

AL1L commented on src/events/types/inventory/InventoryInteractEvent.ts:
```diff
@@ -0,0 +1,46 @@
+  public setCancelled(cancel: boolean): void {
+    this.toJava().setCancelled(cancel);
+  }
+}
```
Add eof newline

AL1L commented on src/events/types/player/PlayerEvent.ts:
```diff
@@ -0,0 +1,16 @@
+  }
+
+  /**
+   * Gets the player who joined the server
```
wrong docs

AL1L commented on src/inventory/Inventory.ts:
```diff
@@ -191,4 +211,15 @@ export class Inventory implements ToJava {
+                }
+            }
+        }
+    }
```
Probably should remove this as it's out of scope

AL1L commented on src/material/Enchantment.ts:
```diff
@@ -3,7 +3,47 @@ import { ToJava } from "../runtime/ToJava";
+    public static readonly SWEEPING_EDGE: Enchantment = JAVA_ENCHANTMENT.SWEEPING_EDGE;
+    public static readonly THORNS: Enchantment = JAVA_ENCHANTMENT.THORNS;
+    public static readonly VANISHING_CURSE: Enchantment = JAVA_ENCHANTMENT.VANISHING_CURSE;
+    public static readonly WATER_WORKER: Enchantment = JAVA_ENCHANTMENT.WATER_WORKER;
```
@connerdouglass I was just experimenting with this, but not sure if I like it, thoughts?

I'm thinking it should return a JS instance of Enchantment instead

connerdouglass replied:
I agree with you, I think it should return a JS instance. Ideally with the customrealms/core library we don't want to expose any Java values that aren't wrapped

connerdouglass replied:
Hopefully updating this file is quick with multiple cursors!

AL1L committed:
Fix small issues with docs and scope

connerdouglass commented on src/events/types/inventory/InventoryInteractEvent.ts:
```diff
@@ -0,0 +1,46 @@
+  /**
+   * Gets the EventResult of this event.
+   */
+  public getResult(): EventResult {
+    return this.toJava().getResult();
+  }
```
I think it should be this, since Java has a `name()` method on enum values that returns a string

```ts
return this.toJava().getResult().name();
```

AL1L replied:
Well, in the way I did EventResult, as per my comment above, this isnt how it works, but I was thinking about changing it to do that

connerdouglass commented on src/events/types/inventory/InventoryInteractEvent.ts:
```diff
@@ -0,0 +1,46 @@
+  /**
+   * Sets the result of this event.
+   */
+  public setResult(result: EventResult): void {
+    this.toJava().setResult(result);
+  }
```
```ts
const javaResult = Java.resolve('org.bukkit.event.Event.Result').valueOf(result);
this.toJava().setResult(javaResult);
```

Gotta test it out to make sure that works, but I think that's it

AL1L replied:
Same as my comment [here](https://github.com/customrealms/core/pull/7#discussion_r753919838)

connerdouglass commented on src/events/types/inventory/InventoryCloseEvent.ts:
```diff
@@ -0,0 +1,28 @@
+    return 'org.bukkit.event.inventory.InventoryCloseEvent';
+  }
+
+  /**
+   * Which button was pressed to initiate the enchanting.
+   */
```
This doc is for the incorrect method

connerdouglass commented on src/events/types/inventory/InventoryEvent.ts:
```diff
@@ -0,0 +1,27 @@
+   * Gets the view object itself
+   */
+  public getView(): InventoryView {
+    return new InventoryView(this.toJava().getView());
```
Don't worry about it now, but eventually I'll want to make an `InventoryView.fromJava` function instead of using the constructor.

AL1L replied:
Got it, this can be eddited when it's added

connerdouglass commented on src/events/types/inventory/InventoryOpenEvent.ts:
```diff
@@ -0,0 +1,28 @@
+    return 'org.bukkit.event.inventory.InventoryOpenEvent';
+  }
+
+  /**
+   * Which button was pressed to initiate the enchanting.
+   */
```
Same thing as above, this doc comment is incorrect. Not a big deal of course

AL1L committed:
Add missing exports

AL1L committed:
Fix doc

AL1L committed:
Fix another doc

AL1L committed:
Change EventResult to be consistent with the rest of the project

connerdouglass commented on src/events/types/inventory/PrepareItemEnchantEvent.ts:
```diff
@@ -0,0 +1,66 @@
+   * null, if there isn't a enchantment offer at a specific slot. There are 3
+   * slots in the enchantment table available to modify.
+   */
+  getOffers(): EnchantmentOffer[] {
+    return this.toJava().getOffers().map((offer: Java.Value) => offer && EnchantmentOffer.fromJava(offer));
+  }
```
If the offer value is nullable, the return type for this method should be changed to `(EnchantmentOffer | null)[]`.

Also the doc mentions that (in a normal Java plugin) you can edit the array to give the player different offers. I don't think this would work in JavaScript though.

AL1L replied:
Correct about the type mismatch, and right about setting elements in the array wouldn't modify the values in java, I'll work on something to make that work. Otherwise, editing the EnchantmentOffers it gives you would modify those offers.

connerdouglass replied:
Right right, just not editing the array itself. No worries, we'll add that functionality in another PR!

AL1L committed:
Use Enchantment.fromJava in default enchantment values

connerdouglass commented on src/events/types/inventory/PrepareItemCraftEvent.ts:
```diff
@@ -0,0 +1,30 @@
+  /**
+   * Get the recipe that has been formed. If this event was triggered by a tool repair, this will be a temporary shapeless recipe representing the repair.
+   */
+  getRecipe(): Recipe {
+    const recipe = this.toJava().getRecipe();
+    return recipe && Recipe.fromJava(recipe);
+  }
```
If the recipe value can be null, update the return type to `Recipe | null` probably

AL1L committed:
Fix incorrect returns

connerdouglass commented on src/material/ItemStack.ts:
```diff
@@ -10,11 +10,15 @@ export class ItemStack implements ToJava {
         return new ItemStack(_itemstack);
     }
 
+    public static fromJavaNullable(_itemstack: Java.Value): ItemStack | null {
+        return _itemstack == null ? null : this.fromJava(_itemstack);
+    }
```
For dumb JavaScript reasons, you need to update `this.fromJava(_itemstack)` to instead be `ItemStack.fromJava(_itemstack)`

Somewhere else in the code (I forget where), you do something like this:
```ts
someArray.map(ItemStack.fromJavaNullable)
```
and when you do that, `fromJavaNullable` is called with an undefined receiver, so `this` will not work.

AL1L committed:
dont use static this

connerdouglass commented on src/material/Enchantment.ts:
```diff
@@ -3,7 +3,47 @@ import { ToJava } from "../runtime/ToJava";
+const JAVA_ENCHANTMENT = Java.resolve('org.bukkit.enchantments.Enchantment');
+
 export class Enchantment implements ToJava {
+    public static readonly ARROW_DAMAGE: Enchantment = Enchantment.fromJava(JAVA_ENCHANTMENT.ARROW_DAMAGE);
+    public static readonly ARROW_FIRE: Enchantment = Enchantment.fromJava(JAVA_ENCHANTMENT.ARROW_FIRE);
+    public static readonly ARROW_INFINITE: Enchantment = Enchantment.fromJava(JAVA_ENCHANTMENT.ARROW_INFINITE);
+    public static readonly ARROW_KNOCKBACK: Enchantment = Enchantment.fromJava(JAVA_ENCHANTMENT.ARROW_KNOCKBACK);
+    public static readonly BINDING_CURSE: Enchantment = Enchantment.fromJava(JAVA_ENCHANTMENT.BINDING_CURSE);
```
This is better. Might change this a bit in the future just for performance. The Enchantment class eagerly loads all the enchantment types even if they're not used.. But I doubt it's going to be noticeable.

connerdouglass reviewed (approved):
Looks good to me 👍

AL1L committed:
Modify getOffers() doc to more accurately describe implementation
